pkg/controller/projects/deploykeys/controller.go

```diff
@@ -0,0 +1,270 @@
+package deploykeys
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	compare "github.com/google/go-cmp/cmp"
+	errors "github.com/pkg/errors"
+
+	v1alpha1 "github.com/crossplane-contrib/provider-gitlab/apis/projects/v1alpha1"
+	clients "github.com/crossplane-contrib/provider-gitlab/pkg/clients"
+	projects "github.com/crossplane-contrib/provider-gitlab/pkg/clients/projects"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	event "github.com/crossplane/crossplane-runtime/pkg/event"
+	logging "github.com/crossplane/crossplane-runtime/pkg/logging"
+	meta "github.com/crossplane/crossplane-runtime/pkg/meta"
+	managed "github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	resource "github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	kubeCoreV1 "k8s.io/api/core/v1"
+	kubeTypes "k8s.io/apimachinery/pkg/types"
+	kubeController "sigs.k8s.io/controller-runtime"
+	kubeClient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+const (
+	errNotDeployKey = "managed resource is not a Gitlab deploy key custom resource"
+	errNotFound     = "404 project deploy key not found"
+	errGetFail      = "cannot get Gitlab deploy key"
+	errCreateFail   = "cannot create Gitlab deploy key"
+	errUpdateFail   = "cannot update Gitlab deploy key"
+	errDeleteFail   = "cannot delete Gitlab deploy key"
+	errKeyMissing   = "missing key ref value"
+	errIdNotAnInt   = "external-name is not an int"
+)
+
+type external struct {
+	kube   kubeClient.Client
+	client projects.DeployKeyClient
+}
+
+type connector struct {
+	kube              kubeClient.Client
+	newGitlabClientFn func(clientConfig clients.Config) projects.DeployKeyClient
+}
+
+func SetupDeployKey(manager kubeController.Manager, log logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.DeployKeyKind)
+
+	return kubeController.NewControllerManagedBy(manager).
+		Named(name).
+		For(&v1alpha1.DeployKey{}).
+		Complete(buildReconciler(manager, log, name))
+}
+
+func (c *connector) Connect(ctx context.Context, mgd resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mgd.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return nil, errors.New(errNotDeployKey)
+	}
+
+	config, err := clients.GetConfig(ctx, c.kube, cr)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &external{kube: c.kube, client: c.newGitlabClientFn(*config)}, nil
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotDeployKey)
+	}
+
+	id := meta.GetExternalName(cr)
+
+	if id == "" {
+		return managed.ExternalObservation{}, nil
+	}
+
+	dkArray, _, err := e.client.ListProjectDeployKeys(
+		*cr.Spec.ForProvider.ProjectId,
```

Comment:
The referencer will only act if the ref fields are defined. But that is not necessarily the case since ref fields are optional as well. You have to add a custom check.